Tutorial and documentation for config-based connectors (#15027)
airbytehq/airbyte: docs/connector-development/config-based/tutorial/5-incremental-reads.md

```
documentationUrl: https://docs.airbyte.io/integrations/sources/exchangeratesapi
connectionSpecification:
  $schema: http://json-schema.org/draft-07/schema#
  title: exchangeratesapi.io Source Spec
  type: object
  required:
    - start_date
    - access_key
    - base
  additionalProperties: true
  properties:
    start_date:
      type: string
      description: Start getting data from that date.
      pattern: ^[0-9]{4}-[0-9]{2}-[0-9]{2}$
      examples:
        - YYYY-MM-DD
    access_key:
      type: string
      description: >-
        Your API Access Key. See <a
        href="https://exchangeratesapi.io/documentation/">here</a>. The key is
        case sensitive.
      airbyte_secret: true
    base:
      type: string
      description: >-
        ISO reference currency. See <a
        href="https://www.ecb.europa.eu/stats/policy_and_exchange_rates/euro_reference_exchange_rates/html/index.en.html">here</a>.
      examples:
        - EUR
        - USD
```

Then we'll set the `start_date` to last week our connection config in `secrets/config.json`.
The following `echo` command will update your config with a start date set at 7 days prior to today.

```

> echo "{\"access_key\": \"<your_access_key>\", \"start_date\": \"$(date -v -7d '+%Y-%m-%d')\", \"base\": \"USD\"}"  > secrets/config.json

Reviewer: I find it a bit overkill to re-use and `echo`, requiring the user to re-copy paste their access key just to benefit from the date function. They should feel comfortable editing config.json directly.

```

And we'll update the `path` in the connector definition to point to `/{{ config.start_date }}`.
Note that we are setting a default value because the `check` operation does not know the `start_date`. We'll default to hitting `/latest`:

```
retriever:
  requester:
      path:
        type: "InterpolatedString"

>         string: "{{ config.start_date }}"

Reviewer: ```suggestion
        string: "/{{ config.start_date }}"
```
Wondering if a leading `/` is required here

>         default: "/latest"

Reviewer: Changing this section fails for me:
from 
```yaml
  retriever:
    $ref: "*ref(retriever)"
    requester:
      $ref: "*ref(requester)"
      path: "/latest"
```

to 
```yaml 
  retriever:
    requester:
        path:
          type: "InterpolatedString"
          string: "{{ config.start_date }}"
          default: "/latest"
```
Leads to the following error on `read`
```python
Exception: failed to create object of type <class 'airbyte_cdk.sources.declarative.requesters.http_requester.HttpRequester'> because __init__() missing 1 required keyword-only argument: 'url_base'`
```

```

> You can test the connector by executing the `read` operation:

Reviewer: ```suggestion
You can test these changes by executing the `read` operation again:
```

```python main.py read --config secrets/config.json --catalog integration_tests/configured_catalog.json```

By reading the output record, you should see that we read historical data instead of the latest exchange rate.
For example:
> "historical": true, "base": "USD", "date": "2022-07-18"

The connector will now always read data for the start date, which is not exactly what we want.
Instead, we would like to iterate over all the dates between the start_date and today and read data for each day.

We can do this by adding a `DatetimeStreamSlicer` to the connector definition, and update the `path` to point to the stream_slice's `start_date`:
More details on the stream slicers can be found [here](./link-to-stream-slicers.md) <FIXME: need to fix links>

> Let's first define a stream slicer at the top level of the connector definition:

Reviewer: It must be before declared before references to it isn't it?

```
stream_slicer:
  type: "DatetimeStreamSlicer"
  start_datetime:
    datetime: "{{ config.start_date }}"
    datetime_format: "%Y-%m-%d"
  end_datetime:

>     datetime: "{{ now_local() }}"

Reviewer: It could be helpful to explain this will generate slices until the current date.

    datetime_format: "%Y-%m-%d %H:%M:%S.%f"
  step: "1d"
  datetime_format: "%Y-%m-%d"
```

and refer to it in the stream's retriever. Note that we're also setting the `cursor_field` in the stream's `options` because it is used both by the `Stream` and the `StreamSlicer`:

```
rates_stream:
  type: DeclarativeStream

>   options:

Reviewer: ```suggestion
  $options:
```

    name: "rates"
    cursor_field: "date
  primary_key: "date"
  schema_loader:
    ref: "*ref(schema_loader)"
  retriever:
    ref: "*ref(retriever)"
    stream_slicer:

>       ref: "*ref(stream_slicer)"

Reviewer: According to your full example below the retriever should not be declared under the `rates_stream`.

```

And we'll update the path to point to the `stream_slice`'s start_date

```

> requester:

Reviewer: I think this is the stream requester, but reading your example I was replacing the root requester.

```
schema_loader:
  type: JsonSchema
  file_path: "./source_exchange_rates_tutorial/schemas/{{ options.name }}.json"
selector:
  type: RecordSelector
  extractor:
    type: JelloExtractor
    transform: "[_]"
requester:
  type: HttpRequester
  name: "{{ options['name'] }}"
  url_base: "https://api.exchangeratesapi.io/v1/"
  http_method: "GET"
  request_options_provider:
    request_parameters:
      access_key: "{{ config.access_key }}"
      base: "{{ config.base }}"
stream_slicer:
  type: "DatetimeStreamSlicer"
  start_datetime:
    datetime: "{{ config.start_date }}"
    datetime_format: "%Y-%m-%d"
  end_datetime:
    datetime: "{{ now_local() }}"
    datetime_format: "%Y-%m-%d %H:%M:%S.%f"
  step: "1d"
  datetime_format: "%Y-%m-%d"
retriever:
  type: SimpleRetriever
  name: "{{ options['name'] }}"
  primary_key: "{{ options['primary_key'] }}"
  record_selector:
    $ref: "*ref(selector)"
  paginator:
    type: NoPagination
  stream_slicer:
    $ref: "*ref(stream_slicer)"
rates_stream:
  type: DeclarativeStream
  $options:
    name: "rates"
    cursor_field: "date"
  primary_key: "date"
  schema_loader:
    $ref: "*ref(schema_loader)"
  retriever:

>     $ref: "*ref(retriever)"

Reviewer: I think your above examples did not have a `$`

```
{
  "access_key": "<your_access_key>",
  "start_date": "2022-07-26",
  "base": "USD"
}
```

where the start date should be 7 days in the past.

And we'll update the `path` in the connector definition to point to `/{{ config.start_date }}`.
Note that we are setting a default value because the `check` operation does not know the `start_date`. We'll default to hitting `/latest`:

```
rates_stream:
  type: DeclarativeStream
  $options:
    name: "rates"
    stream_cursor_field: "date"
  primary_key: "date"
  schema_loader:
    $ref: "*ref(schema_loader)"
  retriever:
    $ref: "*ref(retriever)"
    requester:
      $ref: "*ref(requester)"
      path: "/exchangerates_data/{{ stream_slice['start_time'] or 'latest' }}"
```

You can test these changes by executing the `read` operation:

```python main.py read --config secrets/config.json --catalog integration_tests/configured_catalog.json```

By reading the output record, you should see that we read historical data instead of the latest exchange rate.
For example:
> "historical": true, "base": "USD", "date": "2022-07-18"

The connector will now always read data for the start date, which is not exactly what we want.

> Instead, we would like to iterate over all the dates between the start_date and today and read data for each day.

Reviewer: ```suggestion
Instead, we would like to iterate over all the dates between the start date and today and read data for each day.
```

More details on the stream slicers can be found [here](./link-to-stream-slicers.md) <FIXME: need to fix links>

Let's first define a stream slicer at the top level of the connector definition:

```
stream_slicer:
  type: "DatetimeStreamSlicer"
  start_datetime:
    datetime: "{{ config['start_date'] }}"
    datetime_format: "%Y-%m-%d"
  end_datetime:
    datetime: "{{ now_local() }}"
    datetime_format: "%Y-%m-%d %H:%M:%S.%f"
  step: "1d"
  datetime_format: "%Y-%m-%d"
  cursor_field: "{{ options['stream_cursor_field'] }}"
```

and refer to it in the stream's retriever.
This will generate slices from the start time until the end time, where each slice is exactly one day.
The start time is defined in the config file, while the end time is defined by the `now_local()` macro, which will evaluate to the current date in the current timezone at runtime. See the section on [string interpolation](../yaml-structure.md#string-interpolation) for more details.

Note that we're also setting the `cursor_field` in the stream's `options` because it is used both by the `Stream` and the `StreamSlicer`:

> ```

Reviewer: ```suggestion
```yaml
```

rates_stream:
  type: DeclarativeStream
  $options:
    name: "rates"
    stream_cursor_field: "date"
```

We'll also update the retriever to user the stream slicer:

> ```

Reviewer: ```suggestion
```yaml
```

> retriever:
>   type: SimpleRetriever
>   $options:
>     url_base: "https://api.apilayer.com" # Only change the url_base field
>   name: "{{ options['name'] }}"
>   primary_key: "{{ options['primary_key'] }}"
>   stream_slicer:
>     $ref: "*ref(stream_slicer)"

Reviewer: ```suggestion
retriever:
  type: SimpleRetriever
  $options:
    url_base: https://api.exchangeratesapi.io/v1/
  name: "{{ options['name'] }}"
  primary_key: "{{ options['primary_key'] }}"
  record_selector:
    $ref: "*ref(selector)"
  paginator:
    type: NoPagination
  stream_slicer:
    $ref: "*ref(stream_slicer)"
```

```

Finally, we'll update the path to point to the `stream_slice`'s start_date

> ```

Reviewer: ```suggestion
```yaml
```

rates_stream:
  type: DeclarativeStream
  $options:
    name: "rates"
    stream_cursor_field: "date"
  primary_key: "date"
  schema_loader:
    $ref: "*ref(schema_loader)"
  retriever:
    $ref: "*ref(retriever)"
    requester:
      $ref: "*ref(requester)"

>       path: "/exchangerates_data/{{ stream_slice['start_time'] or 'latest' }}"

Author: @alafanechere I updated the url_base to point to the latest version of their api (https://api.apilayer.com) so the path has to be `/exchangerates_data/{{ stream_slice['start_time'] or 'latest' }}`

```

The full connector definition should now look like `./source_exchange_rates_tutorial/exchange_rates_tutorial.yaml`:

```

> schema_loader:

Reviewer: My fully working YAML at this stage is the following: 
```yaml 
stream_slicer:
  type: "DatetimeStreamSlicer"
  start_datetime:
    datetime: "{{ config['start_date'] }}"
    datetime_format: "%Y-%m-%d"
  end_datetime:
    datetime: "{{ now_local() }}"
    datetime_format: "%Y-%m-%d %H:%M:%S.%f"
  step: "1d"
  datetime_format: "%Y-%m-%d"
  cursor_field: "{{ options['stream_cursor_field'] }}"
schema_loader:
  type: JsonSchema
  file_path: "./source_exchange_rates_tutorial/schemas/{{ options['name'] }}.json"
selector:
  type: RecordSelector
  extractor:
    type: JelloExtractor
    transform: "[_]"
requester:
  type: HttpRequester
  name: "{{ options['name'] }}"
  http_method: "GET"
  request_options_provider:
    request_parameters:
      base: "{{ config.base }}"
      access_key: "{{ config['access_key'] }}"
retriever:
  type: SimpleRetriever
  $options:
    url_base: https://api.exchangeratesapi.io/v1/
  name: "{{ options['name'] }}"
  primary_key: "{{ options['primary_key'] }}"
  record_selector:
    $ref: "*ref(selector)"
  paginator:
    type: NoPagination
  stream_slicer:
    $ref: "*ref(stream_slicer)"
rates_stream:
  type: DeclarativeStream
  $options:
    name: "rates"
    stream_cursor_field: "date"
  primary_key: "date"
  schema_loader:
    $ref: "*ref(schema_loader)"
  retriever:
    $ref: "*ref(retriever)"
    requester:
      $ref: "*ref(requester)"
      path: "/{{ stream_slice['start_time'] or 'latest' }}"
streams:
  - "*ref(rates_stream)"
check:
  type: CheckStream
  stream_names: ["rates"]
```
